lib/crypto/algorithms/megolm.js

```diff
@@ -319,6 +325,23 @@ MegolmEncryption.prototype.onRoomMembership = function(event, member, oldMembers
 };
 
 /**
+ * handle a new user joining a room
+ *
+ * @param {string} userId   new member
+ */
+MegolmEncryption.prototype._onNewRoomMember = function(userId) {
+    // make sure we have a list of this user's devices. We are happy to use a
+    // cached version here: we assume that if we already have a list of the
+    // user's devices, then we already share an e2e room with them, which means
+    // that they will have announced any new devices via an m.new_device.
```

Comment:
>  during a catchup sync, do we always process to_device before the membership events

Yes we do: that's necessary to make sure we get the megolm keys before we try to decrypt timeline messages.

There is a chance of our cache being stale, but I'm hoping to avoid forcing an update here if possible.